force-app/main/default/classes/ReceiptTriggerHandler.cls

```diff
@@ -0,0 +1,106 @@
+                                                                                                                                                    WHERE Id IN:listCasesIdsFromReceipts)
+                                     FROM Contact ];
+
+        Map<Id, Contact> doctorsMap = new Map<Id, Contact>();
+        List<Id> doctorsDepartmentsIdsList = new List<Id>();
+
+        for (Contact contact_i : doctorsList){
+            doctorsMap.put(contact_i.Id, contact_i);
+            doctorsDepartmentsIdsList.add(contact_i.AccountId);
+        }
+
+        List<Case> cases = [SELECT Id, ContactId, Amount_of_payments__c
+                            FROM Case];
+        List<AggregateResult> allDoctorsTreatmentsList = [SELECT SUM(Amount_of_payments__c)amount, ContactId ctId
+                                                          FROM Case 
+                                                          WHERE ContactId IN:doctorsMap.keySet() AND CreatedDate = THIS_MONTH
+                                                          GROUP BY ContactId];
+
+        List<Contact> doctorIdsToUpdateList = new List<Contact>();
+        //3.1
+
+        for (AggregateResult result_i : allDoctorsTreatmentsList){
+
+
+
+            if (doctorsMap.get((Id) result_i.get('ctId')).Account.Doctor_payments_to_receive_bonus__c < (Decimal) result_i.get('amount')){
+                doctorIdsToUpdateList.add(new Contact(Id = (Id) result_i.get('ctId'), Payments_amount__c = (Decimal) result_i.get('amount')));
+            }
+        }
+        List<Database.SaveResult> updateResults = Database.update (doctorIdsToUpdateList, true);
+
+        for (Database.SaveResult result : updateResults){
+            if (result.isSuccess()){
+                System.debug('Successfully published event.');
+            } else{
+                for (Database.Error error : result.getErrors()){
+                    System.debug(error.getMessage());
+                    System.debug(error.getStatusCode());
+                }
+            }
+        }
+
+        //3.2
+        List<Account> departmentList = [SELECT Id, Name, (SELECT Id, Payments_amount__c, Best_doctor_of_the_month__c
+                                                          FROM Contacts
+                                                          ORDER BY Payments_amount__c DESC)
+                                        FROM Account 
+                                        WHERE Id IN:doctorsDepartmentsIdsList];
+
+        Map<Id, Contact> doctorsToUpdateMap = new Map<Id, Contact>();
+
+        for (Account department_i : departmentList){
+            Id doctorIdWithHighestPayments = department_i.Contacts[0].Id;
+            Decimal highestAmount =department_i.Contacts[0].Payments_amount__c;
+            for (Contact contact_i : department_i.Contacts){
+                doctorsToUpdateMap.put(contact_i.Id, (new Contact(Id = contact_i.Id, Best_doctor_of_the_month__c = false)));
```

Comment:
I don't think you need to do for-loop here and set Best_doctor_of_the_month__c = false
most likely you need to query only Contact that are marked as "Best Doctor" now and compare it with Contacts to whose you updated Payments_amount__c value on line 68